Let read/write reuse the preallocated memory

qwe661234 opened the pull request:
When running nyancat.elf, syscall read/write create about 200 KiB memory. This modification enable syscall read/write to reuse the pre-allocated memory.

jserv commented on src/syscall.c:
```diff
@@ -85,7 +86,25 @@ static void syscall_write(riscv_t *rv)
     /* read the string being printed */
-    uint8_t *tmp = malloc(count);
+
+    while (count > 4096) {
```
Avoid hardcoding `4096`.

jserv commented on src/syscall.c:
```diff
@@ -10,7 +10,7 @@
 #include "state.h"
 #include "utils.h"
-
+#define PRE_ALLOCATED_SIZE 4096
```
Rename to `PREALLOC_SIZE` for short.

jserv commented on src/syscall.c:
```diff
@@ -85,7 +86,26 @@ static void syscall_write(riscv_t *rv)
+            rv_set_reg(rv, rv_reg_a0, written);
+        } else {
+            /* error */
+            rv_set_reg(rv, rv_reg_a0, -1);
```
If an error occurs, shall we escape from loop?

jserv commented on src/syscall.c:
```diff
@@ -85,7 +86,26 @@ static void syscall_write(riscv_t *rv)
+                fwrite(tmp, 1, PREALLOC_SIZE, map_iter_value(&it, FILE *));
+
+            /* return number of bytes written */
+            rv_set_reg(rv, rv_reg_a0, written);
```
Check if the accumulated byte is correct during multiple iterations.

jserv commented on src/syscall.c:
```diff
@@ -10,7 +10,7 @@
 #include "state.h"
 #include "utils.h"
-
+#define PREALLOC_SIZE 4096
```
Add a blank line prior to `#define`.

jserv commented on src/syscall.c:
```diff
@@ -85,7 +88,27 @@ static void syscall_write(riscv_t *rv)
+            rv_set_reg(rv, rv_reg_a0, written);
+            total_write += written;
+        } else
+            /* error */
```
This comment is not meaningful. Remove.

jserv commented on src/syscall.c:
```diff
@@ -97,12 +120,13 @@ static void syscall_write(riscv_t *rv)
         rv_set_reg(rv, rv_reg_a0, written);
-    } else {
+        total_write += written;
+    } else
```
If we can return earlier, there is no need to place `else` statement here.

Commit:
Modify syscall implementation with fixed memory

When running nyancat.elf, syscall read/write create about 200 KiB
memory. This modification enable syscall read/write to reuse the
pre-allocated memory.
